src/reactor/windows.rs

```diff
@@ -30,31 +47,57 @@ impl Registration {
     ///
     /// The provided file descriptor must be valid and not be closed while this object is alive.
     pub(crate) unsafe fn new(f: BorrowedSocket<'_>) -> Self {
-        Self {
-            raw: f.as_raw_socket(),
-        }
+        Self::Socket(f.as_raw_socket())
+    }
+
+    /// Create a new [`Registration`] around a waitable handle.
+    ///
+    /// # Safety
+    ///
+    /// The provided handle must be valid and not be closed while this object is alive.
+    pub(crate) unsafe fn new_waitable(f: BorrowedHandle<'_>) -> Self {
+        Self::Handle(f.as_raw_handle())
     }
 
     /// Registers the object into the reactor.
     #[inline]
     pub(crate) fn add(&self, poller: &Poller, token: usize) -> Result<()> {
         // SAFETY: This object's existence validates the invariants of Poller::add
-        unsafe { poller.add(self.raw, Event::none(token)) }
+        unsafe {
+            match self {
+                Self::Socket(raw) => poller.add(*raw, Event::none(token)),
+                Self::Handle(handle) => {
+                    poller.add_waitable(*handle, Event::none(token), PollMode::Oneshot)
+                }
+            }
+        }
     }
 
     /// Re-registers the object into the reactor.
     #[inline]
     pub(crate) fn modify(&self, poller: &Poller, interest: Event) -> Result<()> {
         // SAFETY: self.raw is a valid file descriptor
-        let fd = unsafe { BorrowedSocket::borrow_raw(self.raw) };
-        poller.modify(fd, interest)
+        match self {
+            Self::Socket(raw) => {
+                poller.modify(unsafe { BorrowedSocket::borrow_raw(*raw) }, interest)
+            }
+            Self::Handle(handle) => poller.modify_waitable(
+                unsafe { BorrowedHandle::borrow_raw(*handle) },
+                interest,
+                PollMode::Oneshot,
+            ),
+        }
     }
 
     /// Deregisters the object from the reactor.
     #[inline]
     pub(crate) fn delete(&self, poller: &Poller) -> Result<()> {
         // SAFETY: self.raw is a valid file descriptor
-        let fd = unsafe { BorrowedSocket::borrow_raw(self.raw) };
-        poller.delete(fd)
+        match self {
+            Self::Socket(raw) => poller.delete(unsafe { BorrowedSocket::borrow_raw(*raw) }),
+            Self::Handle(handle) => {
+                poller.remove_waitable(unsafe { BorrowedHandle::borrow_raw(*handle) })
```

Comment:
poller API nitpick (although not really related to this PR): why are the poller methods for sockets and handles named wildly different?